File: AjaxVectorObjects/ClientScripts/CanvasData.js
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
Type.registerNamespace("Aurigma.GraphicsMill.AjaxControls.VectorObjects");

Aurigma.GraphicsMill.AjaxControls.VectorObjects.CanvasData = function (canvasObject, forService) {
    ///	<summary>This class represents a state of a canvas and allows applying this state to the <see cref="T:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Canvas" /> class instance.</summary>
    /// <constructor>
    /// 	<exclude />
    /// </constructor>
    /// <field name="Layers" type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.LayerCollection" static="true"><summary>The collection of layers associated with this canvas.</summary></field>
    /// <field name="ScreenXDpi" type="Number" static="true"><summary>The screen horizontal resolution.</summary></field>
    /// <field name="ScreenYDpi" type="Number" static="true"><summary>The screen vertical resolution.</summary></field>
    /// <field name="Zoom" type="Number" static="true"><summary>The zoom value.</summary></field>
    /// <field name="ReturnValue" type="String" static="true"><summary>The value returned by a remote scripting method.</summary></field>
    /// <field name="PreviewColorManagementEnabled" type="Boolean" static="true"><summary>The value indicating whether to use color management when displaying images.</summary></field>
    /// <field name="PrintColorManagementEnabled" type="Boolean" static="true"><summary>The value which specifies whether to use color management when rendering the workspace.</summary></field>
    Aurigma.GraphicsMill.AjaxControls.VectorObjects.CanvasData.initializeBase(this);

    if (canvasObject) {
        this.RgbCPId = canvasObject._rgbColorProfileFileId;
        this.CmykCPId = canvasObject._cmykColorProfileFileId;
        this.GrayscaleCPId = canvasObject._grayscaleColorProfileFileId;
        this.PreviewCM = canvasObject.get_previewColorManagementEnabled();
        this.PrintCM = canvasObject.get_printColorManagementEnabled();
        this.PTCS = canvasObject.get_previewTargetColorSpace();
        this.Z = canvasObject.get_zoom();
        this.TDpi = canvasObject.get_targetDpi();
        this.XDpi = canvasObject.get_screenXDpi();
        this.YDpi = canvasObject.get_screenYDpi();

        if (forService === false || forService === undefined) {
            this.WW = canvasObject._workspaceWidth;
            this.WH = canvasObject._workspaceHeight;
            var layers = canvasObject.get_layers();
            this.L = [];
            this.HD = new Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.HistoryData(canvasObject.get_history());
            for (var i = 0; i < layers.get_count() ; i++) {
                this.L.push(new Aurigma.GraphicsMill.AjaxControls.VectorObjects.LayerData(layers.get_item(i)));
            }

            this.CSO = canvasObject.get_canvasClientSideOptions();

            this.ReGC = canvasObject.get_resizeGripColor();
            this.RoGC = canvasObject.get_rotationGripColor();
            this.ReGS = canvasObject.get_resizeGripSize();
            this.RoGS = canvasObject.get_rotationGripSize();
            this.RV = canvasObject.get_returnValue();
            this.SB = canvasObject.get_isSquaredBackground();

            this.MW = canvasObject.get_marginWidth();
            this.MC = canvasObject.get_marginColor();
            this.LRM = canvasObject.get_leftRightMargin();
            this.TBM = canvasObject.get_topBottomMargin();
            this.CM = canvasObject.get_constrainedMarginEnabled();
            this.Tags = canvasObject.get_tags();
            this.LIU = canvasObject.get_loadingImageUrl();
            this.SC = canvasObject.get_selectionColor();
            this.SW = canvasObject.get_selectionWidth();
            this.ReGLC = canvasObject.get_resizeGripLineColor();
            this.RoGLC = canvasObject.get_rotationGripLineColor();
            this.RoGLL = canvasObject.get_rotationGripLineLength();

            this.BGCC = canvasObject.get_buttonGroupCssClass();
            this.SBCC = canvasObject.get_selectButtonCssClass();
            this.EBCC = canvasObject.get_editButtonCssClass();
            this.DBCC = canvasObject.get_doneButtonCssClass();

            this.SBT = canvasObject.get_selectButtonTitle();
            this.EBT = canvasObject.get_editButtonTitle();
            this.DBT = canvasObject.get_doneButtonTitle();

            this.MSE = canvasObject.get_multipleSelectionEnabled();
            this.MMT = canvasObject.get_mouseMoveTimeout();
            this.DS = canvasObject.get_disableSmoothing();
        }
    }
}

Aurigma.GraphicsMill.AjaxControls.VectorObjects.CanvasData.applyState = function (canvasData, canvasObject) {
    /// <summary>Applies the <paramref name="canvasData" /> state to the specified <see cref="T:J:Aurigma.GraphicsMill.AjaxControls.VectorObjects.Canvas" />.</summary>
    /// <param name="canvasData" type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.CanvasData">The state to apply.</param>
    /// <param name="canvasObject" type="Aurigma.GraphicsMill.AjaxControls.VectorObjects.Canvas">Canvas to apply the state to.</param>
    if ((canvasObject) && (canvasData)) {
        canvasObject._initialization = true;
        canvasObject.get_history().set_locked(true);

        canvasObject.set_workspaceSize(canvasData.WW, canvasData.WH);

        canvasObject.set_targetDpi(canvasData.TDpi);

        // Color profiles
        canvasObject._rgbColorProfileFileId = canvasData.RgbCPId;
        canvasObject._cmykColorProfileFileId = canvasData.CmykCPId;
        canvasObject._grayscaleColorProfileFileId = canvasData.GrayscaleCPId;

        //color management
        canvasObject.set_previewColorManagementEnabled(canvasData.PreviewCM);
        canvasObject.set_printColorManagementEnabled(canvasData.PrintCM);
        canvasObject.set_previewTargetColorSpace(canvasData.PTCS);

        var i, j, layer;
        //contain old VObjects and Layers
        var objectsHash = {};
        var putObject = function (o) {
            if (o.get_uniqueId)
                objectsHash[o.get_uniqueId()] = o;
        }

        var ls = canvasObject.get_layers();
        for (i = 0, lscnt = ls.get_count() ; i < lscnt; i++) {
            layer = ls.get_item(i);
            //save Layer
            putObject(layer);
            var objs = layer.get_vObjects();
            for (j = 0, objscnt = objs.get_count() ; j < objscnt; j++) {
                //save VObject
                putObject(objs.get_item(j));
            }
            objs.clear();
        }
        ls.clear();
        var layerData;
        for (var i = 0; i < canvasData.L.length; i++) {
            layerData = canvasData.L[i];
            layer = objectsHash[layerData.ID] || new Aurigma.GraphicsMill.AjaxControls.VectorObjects.Layer();
            //exclude Layer object from hash
            objectsHash[layerData.ID] = null;
            Aurigma.GraphicsMill.AjaxControls.VectorObjects.LayerData.applyState(canvasData.L[i], layer, objectsHash);
            canvasObject.get_layers().add(layer);
        }

        //dispose unused objects
        for (var k in objectsHash) {
            var o = objectsHash[k];
            if (o)
                o.dispose();
        }
        delete objectsHash;

        Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo.HistoryData.applyState(canvasData.HD, canvasObject.get_history());
        canvasObject.get_history().set_locked(false);
        canvasObject._returnValue = canvasData.RV;
        canvasObject._canvasClientSideOptions = canvasData.CSO;
        canvasObject.initialization = true;
        canvasObject.set_resizeGripColor(canvasData.ReGC);
        canvasObject.set_rotationGripColor(canvasData.RoGC);
        canvasObject.set_resizeGripSize(canvasData.ReGS);
        canvasObject.set_rotationGripSize(canvasData.RoGS);
        canvasObject.set_isSquaredBackground(canvasData.SB);
        canvasObject.set_marginColor(canvasData.MC);
        canvasObject.set_marginWidth(canvasData.MW);
        canvasObject.set_leftRightMargin(canvasData.LRM);
        canvasObject.set_topBottomMargin(canvasData.TBM);
        canvasObject.set_constrainedMarginEnabled(canvasData.CM);
        canvasObject.set_tags(canvasData.Tags);
        canvasObject.set_loadingImageUrl(canvasData.LIU);
        canvasObject.set_selectionColor(canvasData.SC);
        canvasObject.set_selectionWidth(canvasData.SW);
        canvasObject.set_resizeGripLineColor(canvasData.ReGLC);
        canvasObject.set_rotationGripLineColor(canvasData.RoGLC);
        canvasObject.set_rotationGripLineLength(canvasData.RoGLL);

        canvasObject.set_buttonGroupCssClass(canvasData.BGCC);
        canvasObject.set_selectButtonCssClass(canvasData.SBCC);
        canvasObject.set_editButtonCssClass(canvasData.EBCC);
        canvasObject.set_doneButtonCssClass(canvasData.DBCC);

        canvasObject.set_selectButtonTitle(canvasData.SBT);
        canvasObject.set_editButtonTitle(canvasData.EBT);
        canvasObject.set_doneButtonTitle(canvasData.DBT);

        canvasObject.set_multipleSelectionEnabled(canvasData.MSE);
        canvasObject.set_mouseMoveTimeout(canvasData.MMT);
        canvasObject.set_disableSmoothing(canvasData.DS);

        canvasObject._initialization = false;
    }
};

Aurigma.GraphicsMill.AjaxControls.VectorObjects.CanvasData.prototype = {
    L: [],

    XDpi: 72.0,
    YDpi: 72.0,
    Z: 1,
    RV: null,
    PreviewCM: false,
    PrintCM: false
}

Aurigma.GraphicsMill.AjaxControls.VectorObjects.CanvasData.registerClass("Aurigma.GraphicsMill.AjaxControls.VectorObjects.CanvasData", null);

File: AjaxVectorObjects/Canvas/CanvasData.cs
// Copyright (c) 2018 Aurigma Inc. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
//
using Aurigma.GraphicsMill.AjaxControls.VectorObjects.Math;
using Aurigma.GraphicsMill.AjaxControls.VectorObjects.RedoUndo;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using System;
using System.Collections.Generic;

namespace Aurigma.GraphicsMill.AjaxControls.VectorObjects
{
    /// <summary>
    /// Data object to transfer canvas state between client and server
    /// and to serialize canvas state. Short property names used to
    /// optimize length of JSON serializaed data.
    /// </summary>
    [Serializable]
    internal class CanvasData
    {
        public CanvasData()
        {
            CurLI = -1;
            CurVOI = -1;
            HD = new HistoryData();
            L = new List<LayerData>();
            RgbCPId = null;
            CmykCPId = null;
            GrayscaleCPId = null;
            WH = WW = 72.0F;

            Z = 1.0F;
            XDpi = 72.0F;
            YDpi = 72.0F;
            RV = null;
            CSO = new CanvasClientSideOptions();
            SB = true;
            PreviewCM = false;
            PrintCM = false;
            MC = "rgba(100,255,100,0)";
            MW = 0F;
            M = UnitConverter.ConvertUnitsToUnits(300.0F, 0.25F, Unit.Inch, Unit.Point);
            CM = false;

            Tags = new Dictionary<string, object>();
        }

        public CanvasData(Canvas cv)
        {
            HD = new HistoryData(cv.History);
            CurLI = cv.CurrentLayerIndex;
            CurVOI = cv.CurrentVObjectIndex;
            RgbCPId = cv.RgbColorProfileFileId;
            CmykCPId = cv.CmykColorProfileFileId;
            GrayscaleCPId = cv.GrayscaleColorProfileFileId;
            PreviewCM = cv.PreviewColorManagementEnabled;
            PrintCM = cv.PrintColorManagementEnabled;
            PTCS = cv.PreviewTargetColorSpace;
            WW = cv.WorkspaceWidth;
            WH = cv.WorkspaceHeight;
            L = new List<LayerData>();
            for (int i = 0; i < cv.Layers.Count; i++)
            {
                L.Add(new LayerData(cv.Layers[i]));
            }

            Z = cv.Zoom;
            XDpi = cv.ScreenXDpi;
            YDpi = cv.ScreenYDpi;
            TDpi = cv.TargetDpi;
            ReGS = cv.ResizeGripSize;
            ReGC = Common.ConvertToWebColor(cv.ResizeGripColor);
            RoGS = cv.RotationGripSize;
            RoGC = Common.ConvertToWebColor(cv.RotationGripColor);
            CSO = cv.CanvasClientSideOptions;
            RV = cv.ReturnValue;
            SB = cv.IsSquaredBackground;
            MC = Common.ConvertToWebColor(cv.MarginColor);
            MW = cv.MarginWidth;
            LRM = cv.LeftRightMargin;
            TBM = cv.TopBottomMargin;
            CM = cv.ConstrainedMarginEnabled;
            Tags = cv.Tags;
            LIU = cv.LoadingImageUrl;
            SW = cv.SelectionWidth;
            SC = Common.ConvertToWebColor(cv.SelectionColor);
            ReGLC = Common.ConvertToWebColor(cv.ResizeGripLineColor);
            RoGLC = Common.ConvertToWebColor(cv.RotationGripLineColor);
            RoGLL = cv.RotationGripLineLength;

            BGCC = cv.ButtonGroupCssClass;
            SBCC = cv.SelectButtonCssClass;
            EBCC = cv.EditButtonCssClass;
            DBCC = cv.DoneButtonCssClass;
            SBT = cv.SelectButtonTitle;
            EBT = cv.EditButtonTitle;
            DBT = cv.DoneButtonTitle;

            MSE = cv.MultipleSelectionEnabled;
            MMT = cv.MouseMoveTimeout;
            DS = cv.DisableSmoothing;
        }

        public CanvasData(CanvasSlim cv)
        {
            RgbCPId = cv.RgbColorProfileFileId;
            CmykCPId = cv.CmykColorProfileFileId;
            GrayscaleCPId = cv.GrayscaleColorProfileFileId;
            PreviewCM = cv.PreviewColorManagementEnabled;
            PrintCM = cv.PrintColorManagementEnabled;
            PTCS = cv.PreviewTargetColorSpace;
            WW = cv.WorkspaceWidth;
            WH = cv.WorkspaceHeight;
            L = new List<LayerData>();
            for (int i = 0; i < cv.Layers.Count; i++)
            {
                L.Add(new LayerData(cv.Layers[i]));
            }

            Z = cv.Zoom;
            XDpi = cv.ScreenXDpi;
            YDpi = cv.ScreenYDpi;
            TDpi = cv.TargetDpi;
            MSE = cv.MultipleSelectionEnabled;
            MMT = cv.MouseMoveTimeout;
            DS = cv.DisableSmoothing;
            Tags = cv.Tags;
        }

        public void ApplyState(Canvas cv)
        {
            cv.IsInitialized = false;

            cv.History.Locked = true;
            cv.Layers.Clear();
            for (int i = 0; i < L.Count; i++)
            {
                Layer l = new Layer();
                L[i].ApplyState(l);
                cv.Layers.Add(l);
            }
            if (HD != null)
            {
                HD.ApplyState(cv.History);
            }
            cv.CurrentLayerIndex = CurLI;
            cv.CurrentVObjectIndex = CurVOI;
            cv.History.Locked = false;

            cv.RgbColorProfileFileId = RgbCPId;
            cv.CmykColorProfileFileId = CmykCPId;
            cv.GrayscaleColorProfileFileId = GrayscaleCPId;
            cv.PrintColorManagementEnabled = PrintCM;
            cv.PreviewColorManagementEnabled = PreviewCM;
            cv.PreviewTargetColorSpace = PTCS;

            cv.WorkspaceHeight = WH;
            cv.WorkspaceWidth = WW;
            cv.Zoom = Z;
            cv.ScreenXDpi = XDpi;
            cv.ScreenYDpi = YDpi;
            cv.TargetDpi = TDpi;
            cv.ResizeGripColor = ReGC != null ? new RgbColor(Common.ParseWebColor(ReGC)) : cv.ResizeGripColor;
            cv.ResizeGripSize = ReGS;
            cv.RotationGripColor = RoGC != null ? new RgbColor(Common.ParseWebColor(RoGC)) : cv.RotationGripColor;
            cv.RotationGripSize = RoGS;
            cv.CanvasClientSideOptions = CSO;
            cv.ReturnValue = RV;
            cv.IsSquaredBackground = SB;
            cv.MarginColor = MC != null ? new RgbColor(Common.ParseWebColor(MC)) : cv.MarginColor;
            cv.MarginWidth = MW;
            cv.ConstrainedMarginEnabled = CM;
            if (!Utils.EqualsOfFloatNumbers(0, M))
            {
                cv.Margin = M;
            }
            else
            {
                cv.LeftRightMargin = LRM;
                cv.TopBottomMargin = TBM;
            }

            cv.Tags.Clear();

            foreach (var key in Tags.Keys)
                cv.Tags[key] = Tags[key];

            cv.LoadingImageUrl = LIU;
            cv.SelectionColor = SC != null ? new RgbColor(Common.ParseWebColor(SC)) : cv.SelectionColor;
            cv.SelectionWidth = SW;
            cv.ResizeGripLineColor = ReGLC != null ? new RgbColor(Common.ParseWebColor(ReGLC)) : cv.ResizeGripLineColor;
            cv.RotationGripLineColor = RoGLC != null ? new RgbColor(Common.ParseWebColor(RoGLC)) : cv.RotationGripLineColor;
            cv.RotationGripLineLength = RoGLL;

            cv.ButtonGroupCssClass = BGCC;
            cv.SelectButtonCssClass = SBCC;
            cv.EditButtonCssClass = EBCC;
            cv.DoneButtonCssClass = DBCC;
            cv.SelectButtonTitle = SBT;
            cv.EditButtonTitle = EBT;
            cv.DoneButtonTitle = DBT;

            cv.MultipleSelectionEnabled = MSE;
            cv.MouseMoveTimeout = MMT;
            cv.DisableSmoothing = DS;

            cv.IsInitialized = true;
        }

        public void ApplyState(CanvasSlim cv)
        {
            cv.Layers.Clear();
            for (int i = 0; i < L.Count; i++)
            {
                Layer l = new Layer();
                L[i].ApplyState(l);
                cv.Layers.Add(l);
            }

            cv.RgbColorProfileFileId = RgbCPId;
            cv.CmykColorProfileFileId = CmykCPId;
            cv.GrayscaleColorProfileFileId = GrayscaleCPId;
            cv.PrintColorManagementEnabled = PrintCM;
            cv.PreviewColorManagementEnabled = PreviewCM;
            cv.PreviewTargetColorSpace = PTCS;

            cv.WorkspaceHeight = WH;
            cv.WorkspaceWidth = WW;
            cv.MultipleSelectionEnabled = MSE;
            cv.MouseMoveTimeout = MMT;
            cv.Zoom = Z;
            cv.TargetDpi = TDpi;
            cv.ScreenXDpi = XDpi;
            cv.ScreenYDpi = YDpi;
            cv.DisableSmoothing = DS;

            cv.Tags.Clear();

            foreach (var key in Tags.Keys)
                cv.Tags[key] = Tags[key];
        }

        /// <summary>
        /// HistoryData
        /// </summary>
        public HistoryData HD { get; set; }

        /// <summary>
        /// Layers
        /// </summary>
        public List<LayerData> L { get; set; }

        /// <summary>
        /// CanvasClientSideOptions
        /// </summary>
        public CanvasClientSideOptions CSO { get; set; }

        /// <summary>
        /// CurrentLayerIndex
        /// </summary>
        public int CurLI { get; set; }

        /// <summary>
        /// CurrentVObjectIndex
        /// </summary>
        public int CurVOI { get; set; }

        /// <summary>
        /// SelectionWidth
        /// </summary>
        public float SW { get; set; }

        /// <summary>
        /// SelectionColor
        /// </summary>
        public string SC { get; set; }

        /// <summary>
        /// LoadingImageUrl
        /// </summary>
        public string LIU { get; set; }

        /// <summary>
        /// WorkspaceWidth
        /// </summary>
        public float WW { get; set; }

        /// <summary>
        /// WorkspaceHeight
        /// </summary>
        public float WH { get; set; }

        /// <summary>
        /// RgbColorProfileFileId
        /// </summary>
        public string RgbCPId { get; set; }

        /// <summary>
        /// CmykColorProfileFileId
        /// </summary>
        public string CmykCPId { get; set; }

        /// <summary>
        /// GrayscaleColorProfileFileId
        /// </summary>
        public string GrayscaleCPId { get; set; }

        /// <summary>
        /// PreviewColorManagementEnabled
        /// </summary>
        public bool PreviewCM { get; set; }

        /// <summary>
        /// PrintColorManagementEnabled
        /// </summary>
        public bool PrintCM { get; set; }

        /// <summary>
        /// PreviewTargetColorSpace
        /// </summary>
        [JsonConverter(typeof(StringEnumConverter))]
        public ColorSpace? PTCS { get; set; }

        /// <summary>
        /// ConstrainedMarginEnabled
        /// </summary>
        public bool CM { get; set; }

        /// <summary>
        /// MarginColor
        /// </summary>
        public string MC { get; set; }

        /// <summary>
        /// MarginWidth
        /// </summary>
        public float MW { get; set; }

        /// <summary>
        /// Margin
        /// </summary>
        public float M { get; set; }

        /// <summary>
        /// LeftRightMargin
        /// </summary>
        public float LRM { get; set; }

        /// <summary>
        /// TopBottomMargin
        /// </summary>
        public float TBM { get; set; }

        /// <summary>
        /// IsSquaredBackground
        /// </summary>
        public bool SB { get; set; }

        /// <summary>
        /// ReturnValue
        /// </summary>
        public string RV { get; set; }

        /// <summary>
        /// Tags
        /// </summary>
        public Dictionary<string, object> Tags { get; set; }

        /// <summary>
        /// Zoom
        /// </summary>
        public float Z { get; set; }

        /// <summary>
        /// ScreenXDpi
        /// </summary>
        public float XDpi { get; set; }

        /// <summary>
        /// ScreenYDpi
        /// </summary>
        public float YDpi { get; set; }

        /// <summary>
        /// TargetDpi
        /// </summary>
        public float? TDpi { get; set; }

        /// <summary>
        /// ResizeGripSize
        /// </summary>
        public float ReGS { get; set; }

        /// <summary>
        /// ResizeGripColor
        /// </summary>
        public string ReGC { get; set; }

        /// <summary>
        /// ResizeGripLineColor
        /// </summary>
        public string ReGLC { get; set; }

        /// <summary>
        /// RotationGripSize
        /// </summary>
        public float RoGS { get; set; }

        /// <summary>
        /// RotationGripColor
        /// </summary>
        public string RoGC { get; set; }

        /// <summary>
        /// RotationGripLineLength
        /// </summary>
        public float RoGLL { get; set; }

        /// <summary>
        /// RotationGripLineColor
        /// </summary>
        public string RoGLC { get; set; }

        /// <summary>
        /// MultipleSelectionEnabled
        /// </summary>
        public bool MSE { get; set; }

        /// <summary>
        /// MouseMoveTimeout
        /// </summary>
        public int MMT { get; set; }

        /// <summary>
        /// ButtonGroupCssClass
        /// </summary>
        public string BGCC { get; set; }

        /// <summary>
        /// SelectButtonCssClass
        /// </summary>
        public string SBCC { get; set; }

        /// <summary>
        /// EditButtonCssClass
        /// </summary>
        public string EBCC { get; set; }

        /// <summary>
        /// DoneButtonCssClass
        /// </summary>
        public string DBCC { get; set; }

        /// <summary>
        /// SelectButtonTitle
        /// </summary>
        public string SBT { get; set; }

        /// <summary>
        /// EditButtonTitle
        /// </summary>
        public string EBT { get; set; }

        /// <summary>
        /// DoneButtonTitle
        /// </summary>
        public string DBT { get; set; }

        /// <summary>
        /// DisableSmoothing
        /// </summary
        public bool DS { get; set; }
    }
}